types/wlr_output.c

```diff
@@ -288,170 +368,107 @@ static void handle_cursor_surface_commit(struct wl_listener *listener,
 	}
 }
 
-static void handle_cursor_surface_destroy(struct wl_listener *listener,
+static void output_cursor_handle_destroy(struct wl_listener *listener,
 		void *data) {
-	struct wlr_output *output = wl_container_of(listener, output,
-		cursor.surface_destroy);
-
-	wl_list_remove(&output->cursor.surface_commit.link);
-	wl_list_remove(&output->cursor.surface_destroy.link);
-	output->cursor.surface = NULL;
+	struct wlr_output_cursor *cursor = wl_container_of(listener, cursor,
+		surface_destroy);
+	output_cursor_reset(cursor);
 }
 
-void wlr_output_set_cursor_surface(struct wlr_output *output,
+void wlr_output_cursor_set_surface(struct wlr_output_cursor *cursor,
```

Comment:
What if I call this on a cursor that already has an image, or call set_image on a cursor that already has a surface? Is that case handled?

Reply:
Yeah. These things happen right now in rootston because compositor cursors are using `set_image` and client cursors are using `set_surface`.

* If you call `set_surface` on a cursor that has an image, the `wlr_output_cursor.surface` has priority over the image. The `texture` and the `renderer` are not freed because they may be reused by a `set_image` call.
* If you call `set_image` on a cursor that has a surface, the `output_cursor_reset` function cleanups the surface.